pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to apply adjoint jacobian differentiation"""
+
+import warnings
+import numpy as np
+
+import pennylane as qml
+
+from pennylane.operation import operation_derivative
+from pennylane.tape import QuantumTape
+from pennylane.measurements import ExpectationMP
+
+from .apply_operation import apply_operation
+
+from .initialize_state import create_initial_state
+
+# pylint: disable=protected-access, too-many-branches
+
+
+def _dot_product_real(bra, ket, num_wires):
+    """Helper for calculating the inner product for adjoint differentiation."""
+    # broadcasted inner product not summing over first dimension of the bra tensor
+    sum_axes = tuple(range(1, num_wires + 1))
+    return qml.math.real(qml.math.sum(qml.math.conj(bra) * ket, axis=sum_axes))
+
+
+def adjoint_jacobian(
+    tape: QuantumTape, prep_operation: qml.operation.StatePrep = None
+):  # pylint: disable=too-many-statements
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * As it requires knowledge of the statevector, only statevector simulator devices can be
+          used.
+
+        * Only expectation values are supported as measurements.
+
+        * Does not work for parametrized observables like :class:`~.Hamiltonian`,
+          :class:`~.Sum` or :class:`~.Hermitian`.
```

Comment:
Maybe instead say "Cannot differentiate with respect to observables".  We can use `Sum` or `Hermitian`, we just can't differentiate with respect to their parameters.

We should also mention that the observables need a matrix.